react_perf.md

```diff
@@ -221,50 +341,168 @@ Note: This alone does not fix perf problem.
+
+```diff
+  <!DOCTYPE html>
+  <html lang="en">
+    <head>
+      <!-- ... -->
+    </head>
+    <body>
+      <noscript>This site only works with JavaScript enabled. :(</noscript>
+      <div id="root"></div>
++     <script>
++       if ('serviceWorker' in navigator) {
++         window.addEventListener('load', () => {
++           navigator.serviceWorker.register('/service-worker.js');
++         });
++       }
++     </script>
+    </body>
+  </html>
+```
+
+---
+
+### Something’s not right...
+
+We’re precaching an awful lot. <!-- .element: class="fragment" -->
+
+Note: Check the network tab to see how much is being precached.
+
+---
+
+## Ignore vendor bundles
+
+```diff
+  new GenerateSW({
+    cacheId: 'shame-dev',
+    clientsClaim: true,
+    skipWaiting: true,
++   exclude: [/vendor/],
+  })
+```
+
+<small class="fragment">But we _do_ want to cache them if they’re used...</small>
+
+---
+
+## Set up runtime caching for vendor bundles
+
+```diff
+  new GenerateSW({
+    cacheId: 'shame-dev',
+    clientsClaim: true,
+    skipWaiting: true,
+    exclude: [/vendor/],
++   runtimeCaching: [
++     {
++       urlPattern: /vendor/,
++       handler: 'CacheFirst'
++     },
++   ]
+  })
+```
+
+---
+
+## Add runtime caching for external resources
+
+```diff
+    runtimeCaching: [
+      {
+        urlPattern: /vendor/,
+        handler: 'CacheFirst'
+      },
++     {
++       urlPattern: new RegExp('^https://fonts.googleapis.com/'),
++       handler: 'StaleWhileRevalidate',
++       options: {
++         cacheableResponse: {
++           statuses: [0, 200]
++         }
++       }
++     }
+    ]
+```
 
 ---
 
 # Perceived performance
 
-Note: Show things as loaded instead of blocking on the slowest thing.<br>
-Refactor the dashboard to load metrics inside the chart.<br>
-Use Suspense to load the chart.<br>
+- Don’t wait for the slowest load to render things <!-- .element: class="fragment" -->
+
+Note: Refactor the dashboard to load metrics inside the chart.<br>
+Use Suspense to load the chart.
 
 ---
 
-# Optimizing images
+# Optimize images
+
+- That header is... too big <!-- .element: class="fragment" -->
+- Use the right image format <!-- .element: class="fragment" -->
+- Add a srcset <!-- .element: class="fragment" -->
 
 ---
 
-## Want a mini-course in responsive images?
+## Make the images as small as possible
 
-File formats, `srcset`'s, and `<picture>`'s, oh my!
+- Use [SVGO](https://github.com/svg/svgo) to optimize SVGs
+- Use [TinyPNG](https://tinypng.com) to optimize JPGs and PNGs
 
-https://siakaramalegos.github.io/responsive-images-slides/#/
+<small>Protip: Both of these tools are available as webpack loaders.</small>
+
+Note: SVGO is built into the React svg loader we’re using. TinyPNG requires an API key for the webpack loader, which is why we’re not setting it up in the workshop.
 
 ---
 
-# Script execution costs? (third-party scripts)
+## Choose the right image format
+
+- PNG if you need transparency or ultra high quality
```

Comment:
A high-resolution JPG is both smaller and equivalent in visual quality to a PNG in web. Basically, don't use PNGs unless you need transparency + photo-realistic and WEBP isn't supported. PNG's should only be fallbacks for WEBP and SVG for web. Instead of individual file-type optimizations, you can use `image-webpack-loader` or a few other similar tools that optimize all image types.

Reply:
You've got a lot more expertise here, so I'm happy to change this slide as you see fit. Do you want to pull all of the above on this slide?